### ISSUE TITLE
IntegerLookup layer performance issue when inited with vocabulary

### ISSUE THREAD

**Issue by jamm1985** [tensorflow/tensorflow#65610]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

2.14.0

### Custom code

No

### OS platform and distribution

Ubuntu 20.04

### Python version

3.9

### Current behavior?

I ran into a performance issue when calling an IntegerLookup layer initialized with a Python list vocabulary. Please see code snippet.

Note that passing a numpy array as a vocabulary does not cause performance issues, but hangs during serialization since only the list in [code](https://github.com/keras-team/keras/blob/68f9af408a1734704746f7e6fa9cfede0d6879d8/keras/layers/preprocessing/index_lookup.py#L839) is checked.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np

vocab = np.arange(0,10_000,1)
data = np.array([[10,20,30,40], [50,60,70,20_000]])

layer_1 = tf.keras.layers.IntegerLookup(vocabulary=vocab.tolist())
%timeit layer_1(data)
# 6.21 ms ± 349 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) # <- that is issue

layer_2 = tf.keras.layers.IntegerLookup(vocabulary=vocab) # pass as numpy
timeit layer_2(data)
# 285 µs ± 6.04 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

layer_3 = tf.keras.layers.IntegerLookup()
layer_3.adapt(vocab.tolist())
%timeit layer_3(data)
# 286 µs ± 1.31 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

```
```


### Relevant log output

_No response_

**Comment by sushreebarsa** [comment]
Hi, @jamm1985!
Could you please have a look at this [gist](https://colab.research.google.com/gist/sushreebarsa/fec0fc1cc1c00c82eced3bc83fc7d035/65610.ipynb) where I could reproduce this output ? Kindly confirm the results?
```
8.22 ms ± 363 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
509 µs ± 91.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
421 µs ± 7.34 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
``` 

Thank you!

**Comment by jamm1985** [comment]
Hi @sushreebarsa!

This is the same code snippet from the bug report. My timings are slightly different as it depends on system load, but I confirm that `tf.keras.layers.IntegerLookup(vocabulary=vocab.tolist())` is an order of magnitude slower.

**Comment by jamm1985** [comment]
@sushreebarsa from your gist:

```
print(tf.__version__)
# 2.15.0
```

this bug also affects version 2.15.

**Comment by SuryanarayanaY** [comment]
Hi @jamm1985 ,

I have checked with tf-nightly and there seems no performance issue. Please check the attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/b615a3abfd92852654817695db057ecb/65610_tf-nightly.ipynb) and confirm.

**Comment by jamm1985** [comment]
@SuryanarayanaY 

I confirm that this bug does not affect tf-nightly.

**Comment by SuryanarayanaY** [comment]
Hi @jamm1985 ,

Thanks for confirmation. Since the issue resolved in nightly can we mark this as closed? Please feel free to close the issue. Thanks!

**Comment by jamm1985** [comment]
Hi @SuryanarayanaY :

I hope this bug can be fixed in versions 2.14 and 1.15 as these are the most heavily used versions in production right now.

**Comment by tilakrayal** [comment]
@jamm1985,
From the [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/b615a3abfd92852654817695db057ecb/65610_tf-nightly.ipynb) I can see that the issue is fixed in nightly which will be added in the upcoming release might be Tensorflow 2.17 in coming days. For adding the fix in the Tensorflow 2.14 might get the delay of the other branch cuts to be added. 

I request to wait for the other Tensorflow release and check if you are facing the same issue. Please feel free to re-open if the fix is not added in upcoming releases. Thank you!
